cuda_warp.hpp

```diff
@@ -121,36 +121,36 @@ class simd_mask<T, simd_abi::cuda_warp<N>> {
   SIMD_CUDA_ALWAYS_INLINE simd_mask() = default;
   SIMD_ALWAYS_INLINE SIMD_HOST_DEVICE static constexpr
   int size() { return N; }
-  SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE
+  SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE
   simd_mask(bool value)
     :m_value(value)
   {}
-  SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE constexpr
+  SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE constexpr
   bool get() const {
     return m_value;
   }
-  SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE
+  SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE
   simd_mask operator||(simd_mask const& other) const {
     return m_value || other.m_value;
   }
-  SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE
+  SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE
   simd_mask operator&&(simd_mask const& other) const {
     return m_value && other.m_value;
   }
-  SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE
+  SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE
   simd_mask operator!() const {
     return !m_value;
   }
 };
 
 template <class T, int N>
-SIMD_CUDA_ALWAYS_INLINE SIMD_DEVICE
+SIMD_CUDA_ALWAYS_INLINE SIMD_HOST_DEVICE
 bool all_of(simd_mask<T, simd_abi::cuda_warp<N>> const& a) {
   return bool(__all_sync(simd_abi::cuda_warp<N>::mask(), int(a.get())));
```

Comment:
I guess you've compiled so that answers this question, but is `__all_sync` a thing that is okay to use in host code?

Reply:
technically we could protect the inside of this with __CUDA_ARCH__. The SIMD_HOST_DEVICE is more about correct visibility of the function declarations which for example for clang may need to be visible. If it actually would get through to full instantiation of the call on the host it would compile time fail, which I think is what we want.

Reply:
I think the issue is that the KOKKOS_LAMBDA() is a __host__ __device__ function, and so then they throw a wobbly if the functions underneath are only __device__